rfq/negotiator.go

```diff
@@ -0,0 +1,424 @@
+func (n *Negotiator) queryBidFromPriceOracle(peer route.Vertex,
+	assetId *asset.ID, assetGroupKey *btcec.PublicKey,
+	assetAmount uint64) (rfqmsg.OutgoingMsg, error) {
+
+	ctx, cancel := n.WithCtxQuitNoTimeout()
+	defer cancel()
+
+	oracleResponse, err := n.cfg.PriceOracle.QueryBidPrice(
+		ctx, assetId, assetGroupKey, assetAmount,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to query price oracle for "+
+			"bid: %w", err)
+	}
+
+	// If the bid price is nil, then we will return the error message
+	// supplied by the price oracle.
+	if oracleResponse.BidPrice == nil {
+		return nil, fmt.Errorf("price oracle returned error: %v",
+			*oracleResponse.Err)
+	}
+
+	// TODO(ffranr): Ensure that the expiryDelay time is valid and
+	//  sufficient.
+
+	request, err := rfqmsg.NewRequest(
+		peer, assetId, assetGroupKey, assetAmount,
+		*oracleResponse.BidPrice,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create quote request "+
+			"message: %w", err)
+	}
+
+	return request, nil
+}
+
+// RequestQuote requests a bid quote (buying an asset) from a peer.
+func (n *Negotiator) RequestQuote(buyOrder BuyOrder) error {
+	// Query the price oracle for a reasonable bid price. We perform this
+	// query and response handling in a separate goroutine in case it is a
+	// remote service and takes a long time to respond.
+	n.Wg.Add(1)
+	go func() {
+		defer n.Wg.Done()
+
+		// Query the price oracle for a bid price.
+		outgoingMsg, err := n.queryBidFromPriceOracle(
+			*buyOrder.Peer, buyOrder.AssetID,
+			buyOrder.AssetGroupKey, buyOrder.MinAssetAmount,
+		)
+		if err != nil {
+			err := fmt.Errorf("negotiator failed to handle price "+
+				"oracle response: %w", err)
+			n.cfg.ErrChan <- err
+			return
+		}
+
+		// Send the response message to the outgoing messages channel.
+		sendSuccess := fn.SendOrQuit(
+			n.cfg.OutgoingMessages, outgoingMsg, n.Quit,
+		)
+		if !sendSuccess {
+			err := fmt.Errorf("negotiator failed to add quote " +
+				"request message to the outgoing messages " +
+				"channel")
+			n.cfg.ErrChan <- err
+			return
+		}
+	}()
+
+	return nil
+}
+
+// queryAskFromPriceOracle queries the price oracle for an asking price. It
+// returns an appropriate outgoing response message which should be sent to the
+// peer.
+func (n *Negotiator) queryAskFromPriceOracle(
+	request rfqmsg.Request) (rfqmsg.OutgoingMsg, error) {
+
+	// Query the price oracle for an asking price.
+	ctx, cancel := n.WithCtxQuitNoTimeout()
+	defer cancel()
+
+	oracleResponse, err := n.cfg.PriceOracle.QueryAskPrice(
+		ctx, request.AssetID, request.AssetGroupKey,
+		request.AssetAmount, &request.BidPrice,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to query price oracle for ask "+
+			"price: %w", err)
+	}
+
+	// If the asking price is nil, then we will return a quote reject
+	// message which contains the error message supplied by the price
+	// oracle.
+	if oracleResponse.AskPrice == nil {
```

Comment:
Same note: re inverted error checking order.

Reply:
Added price oracle error check regardless of `AskPrice` field value.